File: src/pl/swmud/ns/swaedit/gui/NewAreaWidget.java
package pl.swmud.ns.swaedit.gui;

import java.math.BigInteger;

import pl.swmud.ns.swmud._1_0.area.Area;
import pl.swmud.ns.swmud._1_0.area.Head;
import pl.swmud.ns.swmud._1_0.area.ObjectFactory;
import pl.swmud.ns.swmud._1_0.area.Head.Economy;
import pl.swmud.ns.swmud._1_0.area.Head.Ranges;
import pl.swmud.ns.swmud._1_0.area.Head.Vnums;

import com.trolltech.qt.core.Qt;
import com.trolltech.qt.gui.QMessageBox;
import com.trolltech.qt.gui.QWidget;

class NewAreaWidget extends QWidget {

    private Ui_NewAreaWidget ui = new Ui_NewAreaWidget();
    Signal0 areaCreated = new Signal0();

    public NewAreaWidget() {
        ui.setupUi(this);
        SWAEdit.setChildPosition(this);
        setAttribute(Qt.WidgetAttribute.WA_DeleteOnClose);
        setWindowModality(Qt.WindowModality.ApplicationModal);
    }
    
    private Area newArea(BigInteger lvnum, BigInteger uvnum) {
        ObjectFactory of = new ObjectFactory();
        Area area = of.createArea();
        Head head = of.createHead();
        head.setAuthors("");
        head.setBuilders("");
        Economy economy = of.createHeadEconomy();
        economy.setHigh(0);
        economy.setLow(0);
        head.setEconomy(economy);
        Ranges ranges = of.createHeadRanges();
        ranges.setHigh((short)100);
        ranges.setLow((short)1);
        head.setRanges(ranges);
        pl.swmud.ns.swmud._1_0.area.Head.Reset reset = of.createHeadReset();
        reset.setFrequency(120);
        reset.setMessage("");
        head.setReset(reset);
        Vnums vnums = of.createHeadVnums();
        vnums.setLvnum(lvnum);
        vnums.setUvnum(uvnum);
        head.setVnums(vnums);
        head.setFlags(0);
        head.setName("");
        area.setHead(head);
        area.setMobiles(of.createMobiles());
        area.setObjects(of.createObjects());
        area.setRepairs(of.createRepairs());
        area.setResets(of.createResets());
        area.setRooms(of.createRooms());
        area.setShops(of.createShops());
        area.setSpecials(of.createSpecials());
        
        return area;
    }

    void on_cancelButton_clicked() {
        close();
    }

    void on_acceptButton_clicked() {
        int ammount = 100;
        if (ui.vnums50Radio.isChecked()) {
            ammount = 50;
        }
        else if (ui.vnums100Radio.isChecked()) {
            ammount = 100;
        }
        else if (ui.vnums150Radio.isChecked()) {
            ammount = 150;
        }
        else {
            ammount = ui.vnumsOtherBox.value();
        }
        if (ammount < 50 || ammount % 50 != 0) {
            QMessageBox.critical(null, "Invalid Vnum Ammount", "Vnum ammount must be greater than 49 and divisible by 50!");
            return;
        }
        SWAEdit.ref.area = newArea(BigInteger.ONE,BigInteger.valueOf(ammount));
        hide();
        areaCreated.emit();
        close();
    }
    
    @SuppressWarnings("unused")
    private void on_vnumsOtherRadio_toggled(boolean checked) {
        ui.vnumsOtherBox.setEnabled(checked);
    } 
}

File: src/gui/NewAreaWidget.cpp
#include "gui/NewAreaWidget.h"
#include "ui_NewAreaWidget.h"

#include <QMessageBox>

NewAreaWidget::NewAreaWidget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::NewAreaWidget)
{
    ui->setupUi(this);
    setWindowFlag(Qt::Window);
    setAttribute(Qt::WA_DeleteOnClose);
    setWindowModality(Qt::ApplicationModal);

    // 100 is checked by default (from .ui)
    ui->vnumsOtherBox->setEnabled(false);

    connect(ui->vnumsOtherRadio, &QRadioButton::toggled,
            this, &NewAreaWidget::onOtherRadioToggled);
    connect(ui->acceptButton, &QPushButton::clicked,
            this, &NewAreaWidget::onAcceptClicked);
    connect(ui->cancelButton, &QPushButton::clicked,
            this, &NewAreaWidget::onCancelClicked);
}

NewAreaWidget::~NewAreaWidget()
{
    delete ui;
}

void NewAreaWidget::onOtherRadioToggled(bool checked)
{
    ui->vnumsOtherBox->setEnabled(checked);
}

void NewAreaWidget::onAcceptClicked()
{
    int amount = 0;

    if (ui->vnums50Radio->isChecked()) {
        amount = 50;
    } else if (ui->vnums100Radio->isChecked()) {
        amount = 100;
    } else if (ui->vnums150Radio->isChecked()) {
        amount = 150;
    } else if (ui->vnumsOtherRadio->isChecked()) {
        amount = ui->vnumsOtherBox->value();
    }

    // Validate
    if (amount < 50 || (amount % 50) != 0) {
        QMessageBox::critical(this, "Invalid Vnum Ammount",
            "Vnum ammount must be greater than 49 and divisible by 50!");
        return;
    }

    // Create new area with defaults
    Area area;
    area.head.name = "";
    area.head.authors = "";
    area.head.builders = "";
    area.head.security = 0;
    area.head.vnums.lvnum = 1;
    area.head.vnums.uvnum = amount;
    area.head.flags = 0;
    area.head.economy.low = 0;
    area.head.economy.high = 0;
    area.head.reset.frequency = 120;
    area.head.reset.message = "";
    area.head.ranges.low = 1;
    area.head.ranges.high = 100;

    emit areaCreated(area);
    close();
}

void NewAreaWidget::onCancelClicked()
{
    close();
}
